node_buffer.hpp

```diff
@@ -0,0 +1,194 @@
+/*
+The MIT License (MIT)
+
+Copyright (c) 2013-2014 BitTorrent Inc.
+Copyright (c) 2016 Arvid Norberg
+
+author: Arvid Norberg, Steven Siloti
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of
+this software and associated documentation files (the "Software"), to deal in
+the Software without restriction, including without limitation the rights to
+use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+the Software, and to permit persons to whom the Software is furnished to do so,
+subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#pragma once
+
+#include <array>
+#include <algorithm>
+#include "ip_set.hpp"
+#include "mapped_file.hpp" // for mapped_vector
+
+template <typename Address>
+struct node_buffer
+{
+	using address_type = Address;
+	using steady_clock = std::chrono::steady_clock;
+	using node_id_type = std::array<char, 20>;
+
+	// this is the type of each node entry
+	// in the circular buffer
+	struct node_entry_t
+	{
+		node_id_type node_id;
+		typename Address::bytes_type ip;
+		uint16_t port;
+	};
+
+	node_buffer(char const* filename, size_t const capacity)
+		: m_current_max_size(std::min(size_t(99), capacity))
+		, m_capacity(capacity)
+		, m_last_write_loop(steady_clock::now())
+		, m_buffer(filename, capacity)
+	{
+		// if we picked up nodes from the previous session, we need to restore our
+		// set of IPs too (used to disallow duplicates) as well as making sure the
+		// current_max_size and write cursors are set correctly
+		if (m_buffer.size() > 0) {
```

Comment:
this logic is new. Previously just the buffer was restored, but the m_ips unordered_set was not rebuilt, and hence asserts would fire when trying to work with an existing file